### PR TITLE
Use custom Docker image for CI

### PR DESCRIPTION
**Pull request by AltGr**
Work in progress; ~~image at https://gitlab.inria.fr/lgesbert/catala-ci-images/container_registry/1606 (the repository contains the Dockerfile, generation instructions and gitlab CI to automatically build and provide it)~~

EDIT: the image is now fully built locally, but in a way to preserve caching for the dependencies.

### CODE REVIEW

**Comment by AltGr** [comment]
Closes #253 ; the full image will be built from scratch if absent (including z3), but the Docker cache will naturally enter into play on the build host the next time around, as long as the `.opam` files are not modified.

The main plus is reproducibility, but this also guarantees that the dependencies remain up-to-date. Hopefully, these 1hour+ builds for the base image won't be too frequent, though.

File: catala.opam [hunk 1]
```diff
@@ -39,6 +40,11 @@ depends: [
   "cppo" {>= "1"}
   "alcotest" {with-test & >= "1.5.0"}
   "odoc" {with-doc}
+  "ocamlformat" {cataladevmode & = "0.21.0"}
+  "obelisk" {cataladevmode}
+  "conf-npm" {cataladevmode}
```
[the remark this answers is not in the document]
Comment (AltGr) [reply catala.opam:45]: there is an [upcoming feature](https://github.com/ocaml/opam/pull/5016) to have support in opam (like for test/dev dependencies), but in the meantime a custom variable defined using `OPAMVAR_cataladevmode` works.

File: catala.opam [hunk 2]
```diff
@@ -59,3 +65,9 @@ build: [
   ]
 ]
 dev-repo: "git+https://github.com/CatalaLang/catala.git"
+depexts: [
+  ["groff" "colordiff" "latexmk" "python3-pip" "pandoc"]
```
[the remark this answers is not in the document]
Comment (AltGr) [reply catala.opam:69]: I wanted to ask: `groff` gives an output that is a bit different from `man2html`. Will that matter? We should probably check how it integrates into the webpage.

**Comment by AltGr** [comment]
> That looks perfect, could you add some high-level documentation in `INSTALL.md` about which `opam` invocation to use depending on whether you want to use Catala, dev on it or use Z3 ?

Right! The correct invocations are already present in the Makefile though, so sticking to `make dependencies-*` should be fine.
Hopefully that will be just a single `opam` command in the upcoming release; but at the moment we still have to jump through a few hoops. Pinning first is only needed to workaround ocaml/opam#5046 (which is fixed upstream) for example.

File: catala.opam (hunk 2 above)
[the remark this answers is not in the document]
Comment (denismerigoux) [reply catala.opam:69]: I'll test that now

**Pull request merged**